java/ql/src/experimental/Security/CWE/CWE-094/TemplateInjection.qll

```diff
@@ -0,0 +1,247 @@
+      // Template(java.lang.String name, java.io.Reader reader, Configuration cfg)
+      // Template(java.lang.String name, java.io.Reader reader, Configuration cfg, java.lang.String encoding)
+      // Template(java.lang.String name, java.lang.String sourceCode, Configuration cfg)
+      // Template(java.lang.String name, java.lang.String sourceName, java.io.Reader reader, Configuration cfg)
+      // Template(java.lang.String name, java.lang.String sourceName, java.io.Reader reader, Configuration cfg, ParserConfiguration customParserConfiguration, java.lang.String encoding)
+      // Template(java.lang.String name, java.lang.String sourceName, java.io.Reader reader, Configuration cfg, java.lang.String encoding)
+      exists(ConstructorCall cc, Expr e |
+        cc.getConstructor().getDeclaringType() instanceof TypeFreeMarkerTemplate and
+        e = cc.getAnArgument() and
+        e.getType().(RefType).hasQualifiedName("java.io", "Reader") and
+        this.asExpr() = e
+      )
+    }
+  }
+
+  /**
+   * An argument to FreeMarker template engine's `putTemplate` method call is a sink for `Configuration`.
+   */
+  private class FreeMarkerStringTemplateLoaderPutTemplateSink extends Sink {
+    FreeMarkerStringTemplateLoaderPutTemplateSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(1) and
+        ma.getMethod() instanceof MethodFreeMarkerStringTemplateLoaderPutTemplate
+      )
+    }
+  }
+
+  /**
+   * An argument to Pebble template engine's `getLiteralTemplate` method call is as a sink for `Configuration`.
+   */
+  private class PebbleGetTemplateSinkTemplateSink extends Sink {
+    PebbleGetTemplateSinkTemplateSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(_) and
+        ma.getMethod() instanceof MethodPebbleGetTemplate
+      )
+    }
+  }
+
+  /**
+   * An argument to JinJava template engine's `render` or `renderForResult` method call is as a sink for `Configuration`.
+   */
+  private class JinjavaRenderSink extends Sink {
+    JinjavaRenderSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(_) and
+        ma.getMethod() instanceof MethodJinjavaRenderForResult
+        or
+        ma.getMethod() instanceof MethodJinjavaRender
+      )
+    }
+  }
+
+  /**
+   * An argument to ThymeLeaf template engine's `process` method call is as a sink for `Configuration`.
+   */
+  private class ThymeLeafRenderSink extends Sink {
+    ThymeLeafRenderSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(0) and
+        ma.getMethod() instanceof MethodThymeleafProcess
+      )
+    }
+  }
+
+  private class ThymeLeafgetResourceAsStreamSink extends Sink {
+    ThymeLeafgetResourceAsStreamSink() {
+      exists(MethodAccess ma |
+        this.asExpr() = ma.getArgument(0) and
+        ma.getMethod() instanceof MethodThymeleafGetResourceAsStream and
+        exists(ReturnStmt ret |
+          this.asExpr() = ret.getResult() and
```

Comment:
`this` can't be both the result of a `ReturnStmt` and the argument to a `MethodAccess` at the same time

Reply:
I have removed the `ReturnStmt` case. That is handled by the Spring View query.